### PR TITLE
Edit chains to enable rerequest

### PR DESCRIPTION
**Pull request by Bolodya1997**
* rename, ipcontext, PF ethernetcontext on rerequest should be executed in clients net NS
  * add netns chain element to switch net NS for all subsequent chain elements
  * remove PF ethernetcontext from ethernetcontext - we don't actually need it now
  * move VF ethernetcontext, ipcontext out of connectioncontext
* ipcontext shouldn not fail on readding same neighbors/routes on rerequest
  * fix it

### CODE REVIEW

File: pkg/kernel/link.go [hunk 1]
```diff
@@ -85,7 +86,7 @@ func (l *link) AddAddress(ip string) error {
 	}
 
 	// check if address is already assigned
-	current, err := netlink.AddrList(l.link, netlink.FAMILY_ALL)
+	current, err := netlink.AddrList(l.link, unix.AF_UNSPEC) // netlink.FAMILY_ALL
```
Comment (haiodo) [inline pkg/kernel/link.go:89]: Not clear why, could you add comments why this variant?

File: pkg/kernel/networkservice/connectioncontext/ipcontext/server.go [hunk 2]
```diff
@@ -137,10 +119,10 @@ func setIPNeighbors(ipNeighbours []*networkservice.IpNeighbor, link netlink.Link
 		}
 		if err := netlink.NeighAdd(&netlink.Neigh{
 			LinkIndex:    link.Attrs().Index,
-			State:        netlink.NUD_REACHABLE,
+			State:        0x02, // netlink.NUD_REACHABLE
```
Comment (haiodo) [inline pkg/kernel/networkservice/connectioncontext/ipcontext/server.go:122]: Why not constant?

File: pkg/kernel/networkservice/inject/server.go [hunk 3]
```diff
@@ -30,119 +30,114 @@ import (
...
 
-func (a *injectServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+func (s *injectServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	logEntry := log.Entry(ctx).WithField("injectServer", "Request")
+
 	connID := request.GetConnection().GetId()
 	mech := kernel.ToMechanism(request.GetConnection().GetMechanism())
 
-	nsSwitch, err := utils.NewNSSwitch()
+	nsSwitch, clientNetNSHandle, err := initNetNSSwitchAndHandle(mech.GetNetNSURL())
```
Comment (haiodo) [inline pkg/kernel/networkservice/inject/server.go:50]: It internally do lock, it is hard to read, could you please improve this moment. Thanks.

File: pkg/kernel/networkservice/netns/server.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	})
+	if err != nil {
+		return nil, err
+	}
+	return rv.(*empty.Empty), nil
+}
+
+func runInNetNS(mechanism *networkservice.Mechanism, runner func() (interface{}, error)) (interface{}, error) {
+	if mech := kernel.ToMechanism(mechanism); mech != nil {
+		nsSwitch, err := nsswitch.NewNSSwitch()
```
Comment (haiodo) [inline pkg/kernel/networkservice/netns/server.go:63]: Why lock/unlock is not required here?

File: pkg/kernel/link.go (hunk 1 above)
Comment (Bolodya1997) [reply pkg/kernel/link.go]: `netlink` is a bit of linux-specific and all pretty constants such as `netlink.FAMILY_ALL`, `netlink.NUD_REACHABLE` are not available for OSx developer :)

File: pkg/kernel/networkservice/connectioncontext/ipcontext/server.go (hunk 2 above)
Comment (Bolodya1997) [reply pkg/kernel/networkservice/connectioncontext/ipcontext/server.go]: netlink is a bit of linux-specific and all pretty constants such as netlink.FAMILY_ALL, netlink.NUD_REACHABLE are not available for OSx developer :)

File: pkg/kernel/networkservice/inject/server.go (hunk 3 above)
Comment (Bolodya1997) [reply pkg/kernel/networkservice/inject/server.go:50]: Actually there is only a need to wrap GetNS, SetNS with lock, so I have moved all locking stuff to the NSSwitch internals.

File: pkg/kernel/networkservice/netns/server.go (hunk 4 above)
Comment (Bolodya1997) [reply pkg/kernel/networkservice/netns/server.go:63]: No more need locks at all.

File: pkg/kernel/networkservice/inject/server.go [hunk 5]
```diff
@@ -30,119 +30,109 @@ import (
...
 
-	ifaceName := conn.GetMechanism().GetParameters()[kernel.InterfaceNameKey]
-	if ifaceName == "" {
-		return nil, errors.New("Interface name is not found")
+func initNetNSSwitchAndHandle(netNSURL string) (nsSwitch *nsswitch.NSSwitch, clientNetNSHandle netns.NsHandle, err error) {
+	nsSwitch, err = nsswitch.NewNSSwitch()
+	if err != nil {
+		return nil, -1, errors.Wrap(err, "failed to init net NS switch")
 	}
+	defer func() {
```
Comment (haiodo) [inline pkg/kernel/networkservice/inject/server.go:104]: Please remove

Comment (Bolodya1997) [reply pkg/kernel/networkservice/inject/server.go]: Yep

File: pkg/kernel/tools/nsswitch/ns_switch.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	if s.NetNSHandle, err = netns.Get(); err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// NewNSSwitchAndHandle returns a new NSSwitch and a net NS handle from the given URL
+func NewNSSwitchAndHandle(netNSURLString string) (nsSwitch *NSSwitch, netNSHandle netns.NsHandle, err error) {
+	netNSURL, err := url.Parse(netNSURLString)
```
Comment (haiodo) [inline pkg/kernel/tools/nsswitch/ns_switch.go:49]: I would prefer to see a separate function to instantiate netNSHandle

Comment (Bolodya1997) [reply pkg/kernel/tools/nsswitch/ns_switch.go]: Done

File: pkg/kernel/tools/nshandle/ns_handle.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+	return handle, nil
+}
+
+// RunIn runs runner in the given net NS
+func RunIn(current, target netns.NsHandle, runner func() error) error {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	switch curr, err := netns.Get(); {
```
Comment (haiodo) [inline pkg/kernel/tools/nshandle/ns_handle.go:57]: Should curr.Close() be here as well?

Comment (Bolodya1997) [reply pkg/kernel/tools/nshandle/ns_handle.go]: Yep, fixed.

**Review by haiodo — APPROVED** [review]

**Pull request merged**